pint/toa.py

```diff
@@ -565,52 +569,75 @@ def get_obss(self):
             return self.table['obs']
 
     def get_pulse_numbers(self):
-        """Return a numpy array of the pulse numbers for each TOA"""
+        """Return a numpy array of the pulse numbers for each TOA if they exist"""
         if hasattr(self, "toas"):
             try:
                 return np.array([t.flags['pn'] for t in self.toas]) * u.cycle
             except:
-                log.warning('No pulse numbers for TOAs')
+                log.warning('Pulse numbers do not exist for all TOAs')
                 return None
         else:
             if 'pn' in self.table['flags'][0]:
-                if 'pn' in self.table.colnames:
+                if 'pulse_number' in self.table.colnames:
                     log.error('Pulse number cannot be both a column and TOA flag')
                     raise Exception('Pulse number cannot be both a column and a TOA flag')
                 return np.array(flags['pn'] for flags in self.table['flags']) * u.cycle
-            elif 'pn' in self.table.colnames:
-                return self.table['pn']
+            elif 'pulse_number' in self.table.colnames:
+                return self.table['pulse_number']
             else:
                 log.warning('No pulse numbers for TOAs')
                 return None
 
-
     def get_flags(self):
         """Return a numpy array of the TOA flags"""
         if hasattr(self, "toas"):
             return numpy.array([t.flags for t in self.toas])
         else:
             return self.table['flags']
-
-    def get_flag_value(self, flag, fill_value=None):
-        """Get the request TOA flag values.
-
-           Parameter
-           ---------
-           flag_name: str
-               The request flag name.
-
-           Return
-           ------
-           A list of flag values from each TOA. If the TOA does not have
-           the flag, it will fill up with the fill_value. 
-        """
-        result = []
-        for flags in self.table['flags']:
-            val = flags.get(flag, fill_value)
-            result.append(val)
-        return result
-
+    
+    def get_groups(self, gap_limit=0.0833):
+        '''flag toas within the gap limit (default 2h = 0.0833d) of each other as the same group'''
+        if hasattr(self, "toas") or gap_limit != 0.0833:
+            gap_limit *= u.d
+            mjd_dict = OrderedDict()
+            for i in np.arange(len(self.get_mjds())):
+                mjd_dict[i]=self.get_mjds()[i].value
+            sorted_mjd_list = sorted(mjd_dict.items(), key=lambda kv:(kv[1], kv[0]))
+            indexes = [a[0] for a in sorted_mjd_list]
+            mjds = [a[1] for a in sorted_mjd_list]
+            gaps = np.diff(mjds)
+            lengths = []
+            count = 0
+            for i in range(len(gaps)):
+                if gaps[i]*u.d < gap_limit:
```

Comment:
Stick with Quantity throughout: gap_limit should be input as a Quantity, don't take `.value` of the get_mjds result, and don't multiply units back in. More generally, don't ever use .value without first .to() the units you want.

Reply:
TODO added